File: js/builder/clip.ts
import { AnimationClipData, KeyframeData } from "./types.js";
import { Keyframe } from "./keyframe.js";

export class Clip {
  public id: string;
  private _duration = 1000;
  private _iterations = 1;
  private _keyframes: KeyframeData[] = [];
  private _metadata?: Record<string, any>;

  constructor(id: string) {
    this.id = id;
  }

  public duration(d: number): this {
    this._duration = d;
    return this;
  }

  public metadata(data: Record<string, any>): this {
    this._metadata = { ...this._metadata, ...data };
    return this;
  }

  public iterations(i: number): this {
    this._iterations = i;
    return this;
  }

  public addKeyframe(kf: Keyframe | KeyframeData): this {
    if (kf instanceof Keyframe) {
      this._keyframes.push(kf.build());
    } else {
      this._keyframes.push(kf);
    }
    return this;
  }

  public build(): AnimationClipData {
    return {
      id: this.id,
      duration: this._duration,
      iterations: this._iterations,
      keyframes: [...this._keyframes],
      ...(this._metadata ? { metadata: { ...this._metadata } } : {}),
    };
  }
}

File: starter-kits/motion-studio/rust/keyframe-engine/src/clip.rs
use crate::interpolator::interpolate_keyframes;
use crate::types::{AnimationClipData, TransformData};

#[derive(Debug, Clone)]
pub struct AnimationClip {
    pub data: AnimationClipData,
    pub is_inflated: bool,
}

impl AnimationClip {
    pub fn new(data: AnimationClipData) -> Self {
        let mut clip = Self {
            data,
            is_inflated: false,
        };
        clip.inflate();
        clip
    }

    pub fn inflate(&mut self) {
        if self.is_inflated {
            return;
        }
        // Ensure keyframes are sorted by time
        self.data.keyframes.sort_by(|a, b| a.time.partial_cmp(&b.time).unwrap());
        self.is_inflated = true;
    }

    pub fn evaluate(&self, local_time: f64) -> (TransformData, f32) {
        if self.data.keyframes.is_empty() {
            return (TransformData::default(), 1.0);
        }

        if self.data.keyframes.len() == 1 {
            let kf = &self.data.keyframes[0];
            return (kf.transform.clone(), kf.opacity);
        }

        // Clamp or calculate progress according to duration and iterations
        let effective_time = if self.data.duration > 0.0 {
            if self.data.iterations.is_infinite() {
                local_time % self.data.duration
            } else if local_time >= self.data.duration * self.data.iterations {
                self.data.duration
            } else {
                local_time % self.data.duration
            }
        } else {
            0.0
        };

        // Find keyframe interval
        if effective_time <= self.data.keyframes[0].time {
            let kf = &self.data.keyframes[0];
            return (kf.transform.clone(), kf.opacity);
        }

        let last_idx = self.data.keyframes.len() - 1;
        if effective_time >= self.data.keyframes[last_idx].time {
            let kf = &self.data.keyframes[last_idx];
            return (kf.transform.clone(), kf.opacity);
        }

        for i in 0..last_idx {
            let kf_curr = &self.data.keyframes[i];
            let kf_next = &self.data.keyframes[i + 1];
            if effective_time >= kf_curr.time && effective_time <= kf_next.time {
                return interpolate_keyframes(kf_curr, kf_next, effective_time);
            }
        }

        let kf = &self.data.keyframes[last_idx];
        (kf.transform.clone(), kf.opacity)
    }
}
